An Issue for stringArray in smart contract when compile

Refer to https://github.com/neo-project/neo-devpack-dotnet/pull/187#discussion_r394673272.
`static readonly string[] SupportedStandards = new string[] { "NEP-5", "NEP-10" };` 
This kind of style can cause compilation error.
```
2>Find entrypoint:System.Object Template.NEP5.CSharp.NEP5::Main(System.String,System.Object[])
2>EXEC : Convert error : System.Exception: sth error
2>   at Neo.Compiler.MSIL.ModuleConverter.LinkCode(String main) in D:\Master-smoking\2020-03-20\neo-devpack-dotnet\src\Neo.Compiler.MSIL\MSIL\Converter.cs:line 359
2>   at Neo.Compiler.MSIL.ModuleConverter.Convert(ILModule _in, ConvOption option) in D:\Master-smoking\2020-03-20\neo-devpack-dotnet\src\Neo.Compiler.MSIL\MSIL\Converter.cs:line 200
2>   at Neo.Compiler.Program.Compile(Options options) in D:\Master-smoking\2020-03-20\neo-devpack-dotnet\src\Neo.Compiler.MSIL\Program.cs:line 159
2>D:\Master-smoking\2020-03-20\neo-devpack-dotnet\templates\Template.NEP5.CSharp\Template.NEP5.CSharp.csproj(12,5): error MSB3073: The command "neon -f "D:\Master-smoking\2020-03-20\neo-devpack-dotnet\templates\Template.NEP5.CSharp\bin\Debug\netstandard2.1\Template.NEP5.CSharp.dll" -o" exited with code -1.
2>Done building project "Template.NEP5.CSharp.csproj" -- FAILED.
```

--- a/src/Neo.Compiler.MSIL/MSIL/CctorSubVM.cs
+++ b/src/Neo.Compiler.MSIL/MSIL/CctorSubVM.cs
@@ -32,7 +32,7 @@ namespace Neo.Compiler.MSIL
                 Boolean v = (Boolean)src;
                 return v;
             }
-            else if(src.GetType() == typeof(string[]))
+            else if (src.GetType() == typeof(string[]))
             {
                 string[] strArrays = (string[])src;
                 return strArrays;

--- a/src/Neo.Compiler.MSIL/MSIL/CctorSubVM.cs
+++ b/src/Neo.Compiler.MSIL/MSIL/CctorSubVM.cs
@@ -32,7 +32,12 @@ namespace Neo.Compiler.MSIL
                 Boolean v = (Boolean)src;
                 return v;
             }
-            else
+            else if(src.GetType() == typeof(string[]))
+            {
+                string[] strArrays = (string[])src;
+                return strArrays;
+            }
+            else // TODO support more types
             {
                 return null;
             }
@@ -97,6 +102,13 @@ namespace Neo.Compiler.MSIL
                                 var count = (int)calcStack.Pop();
                                 if (count > MaxArraySize) throw new ArgumentException("MaxArraySize found");
                                 byte[] data = new byte[count];
+                                calcStack.Push(data);
+                            }
+                            else if (src.tokenType == "System.String")
+                            {
+                                var count = (int)calcStack.Pop();
+                                if (count > MaxArraySize) throw new ArgumentException("MaxArraySize found");
+                                string[] data = new string[count];
                                 calcStack.Push(data);
                             }
                             else
@@ -215,6 +227,23 @@ namespace Neo.Compiler.MSIL
                             // field.DeclaringType.FullName + "::" + field.Name;
                         }
                         break;
+                    case CodeEx.Stelem_Ref:
+                        {
+                            var refValue = calcStack.Pop();
+                            if (refValue is string) // Currently, we only support string ref
+                            {
+                                var strValue = (string)refValue;
+                                var index = (int)calcStack.Pop();
+                                var array = calcStack.Pop() as string[];
+                                if (array is null)
+                                {
+                                    constValue = false;
+                                    break;
+                                }
+                                array[index] = strValue;
+                            }
+                            break;
+                        }
                     case CodeEx.Stelem_I1:
                         {
                             var v = (byte)(int)calcStack.Pop();

--- a/src/Neo.Compiler.MSIL/MSIL/Conv_Common.cs
+++ b/src/Neo.Compiler.MSIL/MSIL/Conv_Common.cs
@@ -145,6 +145,17 @@ namespace Neo.Compiler.MSIL
             _ConvertPushDataArray(data, src, to);
         }
 
+        private void _ConvertPushStringArray(string[] strArray, OpCode src, NeoMethod to)
+        {
+            for (int i = strArray.Length - 1; i >= 0; i--)
+            {
+                var str = strArray[i];
+                _ConvertPushString(str, src, to);
+            }
+            _ConvertPushNumber(strArray.Length, src, to);
+            _Insert1(VM.OpCode.PACK, "", to);
+        }
+
         private int _ConvertPushI8WithConv(ILMethod from, long i, OpCode src, NeoMethod to)
         {
             var next = from.GetNextCodeAddr(src.addr);
@@ -286,6 +297,10 @@ namespace Neo.Compiler.MSIL
                     else if (_src is BigInteger bisrc)
                     {
                         _ConvertPushNumber(bisrc, null, to);
+                    }
+                    else if (_src is string[] strArray)
+                    {
+                        _ConvertPushStringArray(strArray, null, to);
                     }
                     else
                     {

--- a/src/Neo.Compiler.MSIL/MSIL/Converter.cs
+++ b/src/Neo.Compiler.MSIL/MSIL/Converter.cs
@@ -1057,6 +1057,10 @@ namespace Neo.Compiler.MSIL
                             {
                                 _ConvertPushNumber(bisrc, src, to);
                             }
+                            else if (_src is string[] strArray)
+                            {
+                                _ConvertPushStringArray(strArray, src, to);
+                            }
                             else
                             {
                                 throw new Exception("not support type Ldsfld\r\n   in: " + to.name + "\r\n");

--- a/tests/Neo.Compiler.MSIL.UnitTests/TestClasses/Contract_Array.cs
+++ b/tests/Neo.Compiler.MSIL.UnitTests/TestClasses/Contract_Array.cs
@@ -72,5 +72,12 @@ namespace Neo.Compiler.MSIL.UnitTests.TestClasses
         {
             return Owner();
         }
+
+        static readonly string[] SupportedStandards = new string[] { "NEP-5", "NEP-10" };
+
+        public static object TestSupportedStandards()
+        {
+            return SupportedStandards;
+        }
     }
 }

--- a/tests/Neo.Compiler.MSIL.UnitTests/UnitTest_Array.cs
+++ b/tests/Neo.Compiler.MSIL.UnitTests/UnitTest_Array.cs
@@ -98,5 +98,19 @@ namespace Neo.Compiler.MSIL.UnitTests
             ByteArray test = new byte[] { 0xf6, 0x64, 0x43, 0x49, 0x8d, 0x38, 0x78, 0xd3, 0x2b, 0x99, 0x4e, 0x4e, 0x12, 0x83, 0xc6, 0x93, 0x44, 0x21, 0xda, 0xfe };
             Assert.IsTrue(ByteArray.Equals(bts, test));
         }
+
+        [TestMethod]
+        public void Test_StringArray()
+        {
+            var testengine = new TestEngine();
+            testengine.AddEntryScript("./TestClasses/Contract_Array.cs");
+            var result = testengine.ExecuteTestCaseStandard("testSupportedStandards");
+
+            var bts = result.Pop().ConvertTo(StackItemType.Array);
+            var items = bts as VM.Types.Array;
+
+            Assert.AreEqual((ByteArray)"NEP-5", items[0]);
+            Assert.AreEqual((ByteArray)"NEP-10", items[1]);
+        }
     }
 }

